1106-i386-slaunch-Add-basic-platform-support-for-secure-l.patch

```diff
@@ -0,0 +1,338 @@
++
++#define GRUB_EFLAGS_X86_CF	0x00000001 /* Carry Flag */
++#define GRUB_EFLAGS_X86_PF	0x00000004 /* Parity Flag */
++#define GRUB_EFLAGS_X86_AF	0x00000010 /* Auxillary carry Flag */
++#define GRUB_EFLAGS_X86_ZF	0x00000040 /* Zero Flag */
++#define GRUB_EFLAGS_X86_SF	0x00000080 /* Sign Flag */
++#define GRUB_EFLAGS_X86_TF	0x00000100 /* Trap Flag */
++#define GRUB_EFLAGS_X86_IF	0x00000200 /* Interrupt Flag */
++#define GRUB_EFLAGS_X86_DF	0x00000400 /* Direction Flag */
++#define GRUB_EFLAGS_X86_OF	0x00000800 /* Overflow Flag */
++#define GRUB_EFLAGS_X86_IOPL	0x00003000 /* IOPL mask */
++#define GRUB_EFLAGS_X86_NT	0x00004000 /* Nested Task */
++#define GRUB_EFLAGS_X86_RF	0x00010000 /* Resume Flag */
++#define GRUB_EFLAGS_X86_VM	0x00020000 /* Virtual Mode */
++#define GRUB_EFLAGS_X86_AC	0x00040000 /* Alignment Check */
++#define GRUB_EFLAGS_X86_VIF	0x00080000 /* Virtual Interrupt Flag */
++#define GRUB_EFLAGS_X86_VIP	0x00100000 /* Virtual Interrupt Pending */
++#define GRUB_EFLAGS_X86_ID	0x00200000 /* CPUID detection flag */
++
++static inline unsigned long
++grub_read_flags_register(void)
++{
++  unsigned long flags;
++
++#ifdef __x86_64__
++  asm volatile ("pushfq; popq %0" : "=r" (flags));
++#else
++  asm volatile ("pushfl; popl %0" : "=r" (flags));
++#endif
++
++  return flags;
++}
++
++static inline void
++grub_write_flags_register(unsigned long flags)
++{
++#ifdef __x86_64__
++  asm volatile ("pushq %0; popfq" : : "r" (flags));
++#else
++  asm volatile ("pushl %0; popfl" : : "r" (flags));
++#endif
++}
++
++#endif
+diff --git a/include/grub/i386/mmio.h b/include/grub/i386/mmio.h
+new file mode 100644
+index 000000000000..b5bce71ef8d6
+--- /dev/null
++++ b/include/grub/i386/mmio.h
+@@ -0,0 +1,72 @@
++/*
++ *  GRUB  --  GRand Unified Bootloader
++ *  Copyright (C) 2020  Oracle and/or its affiliates.
++ *
++ *  GRUB is free software: you can redistribute it and/or modify
++ *  it under the terms of the GNU General Public License as published by
++ *  the Free Software Foundation, either version 3 of the License, or
++ *  (at your option) any later version.
++ *
++ *  GRUB is distributed in the hope that it will be useful,
++ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
++ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ *  GNU General Public License for more details.
++ *
++ *  You should have received a copy of the GNU General Public License
++ *  along with GRUB.  If not, see <https://www.gnu.org/licenses/>.
++ */
++
++#ifndef GRUB_I386_MMIO_H
++#define GRUB_I386_MMIO_H 1
++
++#include <grub/types.h>
++
++static inline grub_uint8_t
++grub_readb (const grub_addr_t addr)
++{
++  grub_uint8_t val;
++
++  val = (*(volatile grub_uint8_t *) (addr));
++
++  return val;
++}
++
++static inline grub_uint32_t
++grub_readl (const grub_addr_t addr)
++{
++  grub_uint32_t val;
++
++  val = (*(volatile grub_uint32_t *) (addr));
++
++  return val;
++}
++
++static inline grub_uint64_t
++grub_readq (const grub_addr_t addr)
```

Comment:
Please use `read64`/`read32`/`read8` instead of `readq`/`readl`/`readb` (and analgous for writes).  This is much easier to read and remember.